### PR TITLE
[functorch] fix multinomial

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83850
* __->__ #83838

### CODE REVIEW

File: functorch/test/test_vmap.py [hunk 1]
```diff
@@ -3807,6 +3807,17 @@ def _assert_all_slices_unique(self, tensor):
         slices_equal.diagonal().zero_()
         self.assertEqual(slices_equal, torch.zeros_like(slices_equal))
 
+    def _assert_correct_size(self, fn, input_tensor, other_args, batch_dim, batch_size, batched_output):
```
Comment (samdow) [inline functorch/test/test_vmap.py:3810]: A note on this: So the actual reason we didn't catch this error with multinomial is because of an issue in the testing (fixed in L4374). But since I had already written this, I added it to the dropout tests. Also happy to split these into two PRs since the multinomial error was just an error in the testing

Comment (samdow) [reply functorch/test/test_vmap.py]: Moved to another PR

File: functorch/functorch/csrc/BatchRulesRandomness.cpp [hunk 2]
```diff
@@ -227,15 +227,16 @@ Tensor multinomial_batching_rule(const Tensor& self, const int64_t num_samples,
     shapeVec.insert(shapeVec.end(), shape.begin(), shape.end());
     self_value = self_value.expand(shapeVec);
   }
-  if (self_value.dim() == 3 && (self_bdim || randomness == RandomnessType::Different)) {
+  const auto orig_input_dim = self_value.dim();
+  if (orig_input_dim == 3 && (self_bdim || randomness == RandomnessType::Different)) {
```
Comment (zou3519) [inline functorch/functorch/csrc/BatchRulesRandomness.cpp:231]: nit: naming a bit off-putting. `orig_input_dim` -- `self_value` on line 218 may have a different shape from self_value on line 228.

Also, this implementation of `multinomial` is confusing, since we're modifying it anyways it is a good chance to clean it up. The thing that is confusing is that it's hard to tell what actually happens if randomness is Same vs Different and what happens in the 1D or 2D multinomial cases.

It might be clearer if we just separate the cases and use helper functions like `rankWithoutBatchDim` to tell us if we're in the 1D or 2D case and `ensure_has_bdim` instead of the manual expand logic above. Pseudocode:
```cpp
if (randomness == RandomnessType::Different) {
  // 1D cases: S -> BS -> multinomial(BS)
  //          BS -> multinomial(BS)
  // 2d cases: MS -> (BM)S -> multinomial(BS)
  //.         BMS -> (BM)S -> multinomial(BS)
  is_2d_case = rankWithoutBatchDim(self_value, self_bdim);
  if (self_bdim) {
    self_value = ensure_has_bdim(self_value, ...)
    if (is_2d_case) {
        self_value = reshape_dim_into(0, 0, self_value);
    }
  }
  auto out = multinomial(...);
  if (is_2d_case) {
    out = reshape_dim_outof(...);
  }
}
if (randomness == RandomnessType::Same) {
  // 1D case: S -> multinomial(S)
  // 2D case: MS -> multinomial(MS)
  out = multinomial(self_value);
  return out;
}


```

**Review by zou3519 — APPROVED** [review]
Code looks right to me, but took a couple of minutes to wade through. I had a suggestion about how to refactor the code to read better

File: functorch/functorch/csrc/BatchRulesRandomness.cpp (hunk 2 above)
Comment (samdow) [reply functorch/functorch/csrc/BatchRulesRandomness.cpp]: Good call--updated to match
